versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md

```diff
@@ -0,0 +1,460 @@
+---
+title: Migrating from ink! 4.x to 5.0
+slug: /faq/migrating-from-ink-4-to-5
+---
+
+<img src="/img/title/migration-4.x-to-5.0.svg" className="titlePic" />
+
+We've made a couple of breaking changes from ink! 4.x to ink! 5.0.
+On this page we outline how you can migrate existing clients and
+contracts from 4.x to 5.0.
+
+This release addresses the majority of issues raised in [the OpenZeppelin
+security review](https://blog.openzeppelin.com/security-review-ink-cargo-contract).
+In particular, we addressed the proxy selector clashing attack.
+
+You can find the full changelog of the 5.0 release [here](https://github.com/paritytech/ink/blob/master/CHANGELOG.md).
+
+:::caution
+This migration guide only considers your code base! Not your data!
+
+If you have an existing contract on-chain you cannot just
+upgrade the code on-chain, you also have to migrate your data.
+
+The way how ink! 4.0 stores data and reads from it (i.e. the storage
+layout) changes from ink! 3.x to 4.0.
```

Comment:
How is this related to migration from 4.x to 5.0?

Reply:
Thanks, improved it.